### PR TITLE
[MLIR][test] Run Integration tests for CPU using qemu-aarch64

### PR DESCRIPTION
**Pull request by zhaoshiz**
To run integration tests using qemu-aarch64 on x64 host, below flags are added to the cmake command when building mlir/llvm:

      -DMLIR_INCLUDE_INTEGRATION_TESTS=ON \
      -DMLIR_RUN_ARM_SVE_TESTS=ON \
      -DMLIR_RUN_ARM_SME_TESTS=ON \
      -DARM_EMULATOR_EXECUTABLE="<...>/qemu-aarch64" \
      -DARM_EMULATOR_OPTIONS="-L /usr/aarch64-linux-gnu" \
      -DARM_EMULATOR_MLIR_CPU_RUNNER_EXECUTABLE="<llvm_arm64_build_top>/bin/mlir-cpu-runner-arm64" \
      -DARM_EMULATOR_LLI_EXECUTABLE="<llvm_arm64_build_top>/bin/lli" \
      -DARM_EMULATOR_UTILS_LIB_DIR="<llvm_arm64_build_top>/lib"

The last three above are prebuilt on, or cross-built for, an aarch64 host.

Some tests marked with 'UNSUPPORTED: target=aarch64{{.*}}' are still run when configured with ARM_EMULATOR_EXECUTABLE and the default target is not aarch64.
A lit config feature 'mlir_arm_emulator' is added in mlir/test/lit.site.cfg.py.in and to UNSUPPORTED list of such tests.

### CODE REVIEW

File: mlir/test/Integration/Dialect/Linalg/CPU/ArmSME/lit.local.cfg
Comment (MacDue) [inline mlir/test/Integration/Dialect/Linalg/CPU/ArmSME/lit.local.cfg:33]: Can this be moved to a common lit config? I see that these are duplicated for the SVE config too.

File: mlir/test/Integration/Dialect/Linalg/CPU/ArmSVE/lit.local.cfg
Comment (MacDue) [inline mlir/test/Integration/Dialect/Linalg/CPU/ArmSVE/lit.local.cfg:33]: Duplicate configuration.

File: mlir/test/Integration/Dialect/LLVMIR/CPU/lit.local.cfg
Comment (MacDue) [inline mlir/test/Integration/Dialect/LLVMIR/CPU/lit.local.cfg:43]: Why just `-mattr=+neon` here?

Comment (c-rhodes) [reply mlir/test/Integration/Dialect/LLVMIR/CPU/lit.local.cfg:43]: also not clear to me what the rationale here is. I don't believe `arm_emulator_executable` was intended to mean run all the integration tests under emulation, it was intended to support running the integration tests for ISAs like SVE and SME enabled under
- MLIR_RUN_ARM_SVE_TESTS
- MLIR_RUN_ARM_SME_TESTS

where hardware isn't yet widely available. I think we need to mindful here because this will mean these tests are unnecessarily run under emulation on AArch64 hosts and will take longer to run.

**Comment by aartbik** [comment]
LG for sparse changes

File: mlir/test/Integration/Dialect/SparseTensor/CPU/sparse_sum_bf16.mlir
Comment (zhaoshiz) [reply mlir/test/Integration/Dialect/SparseTensor/CPU/sparse_sum_bf16.mlir]: changed to UNSUPPORTED: mlir_arm_emulator

File: mlir/test/Integration/Dialect/SparseTensor/CPU/sparse_tanh.mlir
Comment (zhaoshiz) [reply mlir/test/Integration/Dialect/SparseTensor/CPU/sparse_tanh.mlir]: changed to UNSUPPORTED: mlir_arm_emulator

File: mlir/test/Integration/Dialect/Linalg/CPU/ArmSME/lit.local.cfg
Comment (zhaoshiz) [reply mlir/test/Integration/Dialect/Linalg/CPU/ArmSME/lit.local.cfg]: moved to test/Integration/lit.local.cfg

File: mlir/test/Integration/Dialect/Linalg/CPU/ArmSVE/lit.local.cfg
Comment (zhaoshiz) [reply mlir/test/Integration/Dialect/Linalg/CPU/ArmSVE/lit.local.cfg]: moved to test/Integration/lit.local.cfg

File: mlir/test/Integration/Dialect/LLVMIR/CPU/lit.local.cfg
Comment (zhaoshiz) [reply mlir/test/Integration/Dialect/LLVMIR/CPU/lit.local.cfg]: Reverted changes to test/Integration/Dialect/LLVMIR/CPU/lit.local.cfg and sinked substitutions rule of `%mlir_native_utils_lib_dir` to test/Integration/Dialect/SparseTensor/CPU/lit.local.cfg.

File: mlir/test/Integration/Dialect/SparseTensor/CPU/block.mlir
Comment (MacDue) [inline mlir/test/Integration/Dialect/SparseTensor/CPU/block.mlir:13]: Why can't this use the substitutions you've defined? Same question for the other parse tests.
```suggestion
// DEFINE: %{run_libs_sve} = -shared-libs=%native_mlir_runner_utilst,%native_mlir_c_runner_utils
```

File: mlir/test/Integration/Dialect/SparseTensor/CPU/sparse_sum_bf16.mlir
Comment (MacDue) [inline mlir/test/Integration/Dialect/SparseTensor/CPU/sparse_sum_bf16.mlir:33]: I see now that these tests generally don't compile for AArch64 (unrelated to emulation). So, I think this should be something like `UNSUPPORTED: mlir_arm_emulator, target=aarch64{{.*}}`. 

i.e. supported under emulation and real hardware.

**Review by MacDue — COMMENTED** [review]
I've tried this locally and for my use cases everything still seems to work :slightly_smiling_face:, so if this allows people to work on x86 too I'm happy with the changes. Just a few more comments/questions:

File: mlir/test/Integration/Dialect/SparseTensor/CPU/block.mlir
Comment (zhaoshiz) [reply mlir/test/Integration/Dialect/SparseTensor/CPU/block.mlir]: ah, totally missed this part

File: mlir/test/Integration/Dialect/SparseTensor/CPU/sparse_sum_bf16.mlir
Comment (zhaoshiz) [reply mlir/test/Integration/Dialect/SparseTensor/CPU/sparse_sum_bf16.mlir]: ok, I added back target=aarch64{{.*}}

File: mlir/test/Integration/lit.local.cfg
Comment (banach-space) [inline mlir/test/Integration/lit.local.cfg:111]: Perhaps I'm missing something, but couldn't this block be rewritten as:
```python
if config.arm_emulator_utils_lib_dir:
  base_dir = config.arm_emulator_utils_lib_dir
else:
  base_dir = config.mlir_lib_dir
  
config.substitutions.append(("%native_mlir_runner_utils", base_dir + mlir_runner_utils_file_name))
config.substitutions.append((...)))
```
? This way the `if`/`else` blocks would get nice and small.

File: mlir/test/Integration/lit.local.cfg
Comment (banach-space) [inline mlir/test/Integration/lit.local.cfg:52]: The comment on L44-L52 can be moved above this line:
```python
aarch64_mcr_cmd = configure_aarch64_mcr_cmd()
```

Basically, everything that follows is in preparation for possible emulation. I would also move the block that starts with `if config.arm_emulator_utils_lib_dir:` into a function to match `configure_aarch64_mcr_cmd`.

This will most likely require some updates to the comments, but will lead to a much cleaner config.

File: mlir/test/Integration/Dialect/SparseTensor/CPU/sparse_sum_bf16.mlir
Comment (banach-space) [inline mlir/test/Integration/Dialect/SparseTensor/CPU/sparse_sum_bf16.mlir:33]: Why is `mlir_arm_emulator`? Could you add a note in the summary? Thanks!

File: mlir/test/Integration/lit.local.cfg
Comment (zhaoshiz) [reply mlir/test/Integration/lit.local.cfg:66]: Thanks for the suggestion, I've refactored the substitutions.

File: mlir/test/Integration/Dialect/SparseTensor/CPU/sparse_sum_bf16.mlir
Comment (zhaoshiz) [reply mlir/test/Integration/Dialect/SparseTensor/CPU/sparse_sum_bf16.mlir:33]: I added comments in mlir/test/lit.site.cfg.py.in and the summary.

File: mlir/test/Integration/lit.local.cfg
Comment (banach-space) [inline mlir/test/Integration/lit.local.cfg:69]: I would take this one step further and refactor `configure_aarch64_mcr_cmd` to also run:
```python
config.substitutions.append(("%mcr_aarch64_cmd", aarch64_mcr_cmd))
```

This way both methods would be more consistent and the overall code hygiene would improve :)

**Review by banach-space — APPROVED** [review]
This is a very nice improvement, thank you! LGTM

I've left one more small suggestion, but that's just a nit, so feel free to ignore. Please wait for @MacDue to approve before landing.

Thanks again for contributing! 🙏🏻

File: mlir/test/Integration/Dialect/SparseTensor/CPU/lit.local.cfg
Comment (c-rhodes) [inline mlir/test/Integration/Dialect/SparseTensor/CPU/lit.local.cfg:19]: this is no longer used? (and below at the end of the `else`)

**Review by c-rhodes — APPROVED** [review]
left one minor but otherwise LGTM, thanks for fixing this it's a nice improvement!

**Review by MacDue — APPROVED** [review]
LGTM, this is nice! Please address @c-rhodes's comment first though :smile:

File: mlir/test/Integration/Dialect/SparseTensor/CPU/lit.local.cfg
Comment (zhaoshiz) [reply mlir/test/Integration/Dialect/SparseTensor/CPU/lit.local.cfg]: nice catch. I also updated comments about UNSUPPORTing the tests on Windows.

**Pull request merged**
